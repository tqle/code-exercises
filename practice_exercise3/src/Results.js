import React from "react";
import UserTableRow from "./UserTableRow";
import axios from "axios";

const SAMPLE_USERS_URL = "http://jsonplaceholder.typicode.com/users";

class Results extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoading: false,
      users: [], // initial empty state for users
      error: null
    };
  }

  // Fetch users and map the users state once mounted
  // set state users is the same as
  // users: users - same as just users
  componentDidMount() {
    // User isLoading to hide or show progress indicator
    this.setState({ isLoading: true });

    axios
      .get(SAMPLE_USERS_URL)
      .then(response => {
        // Response success
        const users = response.data;
        this.setState({
          users,
          isLoading: false
        });
      })
      .catch(error => {
        // Catch any request and response errors
        // Catch invalid url requests and response
        // ex. 404
        if (error.response) {
          this.setState({
            error: error.response.status,
            isLoading: false
          });
        }
      });
  }

  render() {
    const { error, users } = this.state;

    // Return error message if request or response fails and has an error message.
    // TODO: Add Error Boundary
    if (this.state.error) {
      // TODO: Refactor error message container to new component.
      return <div className="margin-top margin-left">ERROR: {error}</div>;
    } else {
      // if (this.state.isLoading) {
      //   <h1>Loading...</h1>;
      // } else {
      // test
      // TODO: Refactor sort name by descending order as separate re-usable util function
      const sortNameByDescendingOrder = []
        .concat(users)
        .sort((a, b) => a.name.toUpperCase() < b.name.toUpperCase())
        .map(user => (
          <UserTableRow
            key={user.id}
            id={user.id}
            name={user.name}
            username={user.username}
            email={user.email}
            address={user.address}
          />
        ));
      return (
        <div>
          {/* <pre>
            <code>{JSON.stringify(this.state, null, 4)}</code>
          </pre> */}
          <table className="table container-border neutral-background-color">
            <thead>
              <tr>
                <th className="cell-border text-align-left small-padding-top small-padding-right small-padding-left small-padding-bottom">
                  Name
                </th>
                <th className="cell-border text-align-left small-padding-top small-padding-right small-padding-left small-padding-bottom">
                  UserName
                </th>
                <th className="cell-border text-align-left small-padding-top small-padding-right small-padding-left small-padding-bottom">
                  E-mail
                </th>
                <th className="cell-border text-align-left small-padding-top small-padding-right small-padding-left small-padding-bottom">
                  Address
                </th>
              </tr>
            </thead>
            <tbody>{sortNameByDescendingOrder}</tbody>
          </table>
        </div>
      );
      // end test
      // }
    }
  }
}

export default Results;
