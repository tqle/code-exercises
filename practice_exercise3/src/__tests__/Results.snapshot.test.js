import React from "react";
import { create } from "react-test-renderer";
import Results from "../Results";

/**
 * Test User Table Results snapshot
 */
test("<Results /> renders correctly", () => {
  const user = {
    id: 1,
    name: "Leanne Graham",
    username: "Bret",
    email: "Sincere@april.biz",
    address: {
      street: "Kulas Light",
      suite: "Apt. 556",
      city: "Gwenborough",
      zipcode: "92998-3874",
      geo: {
        lat: "-37.3159",
        lng: "81.1496"
      }
    },
    phone: "1-770-736-8031 x56442",
    website: "hildegard.org",
    company: {
      name: "Romaguera-Crona",
      catchPhrase: "Multi-layered client-server neural-net",
      bs: "harness real-time e-markets"
    }
  };
  const test = create(
    <Results
      key={user.id}
      id={user.id}
      name={user.name}
      username={user.username}
      email={user.email}
      address={user.address}
    />
  );
  expect(test.toJSON()).toMatchSnapshot();
});
