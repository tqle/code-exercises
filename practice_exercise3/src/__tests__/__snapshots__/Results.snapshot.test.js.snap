// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Results /> renders correctly 1`] = `
<div>
  <table
    className="table container-border neutral-background-color"
  >
    <thead>
      <tr>
        <th
          className="cell-border text-align-left small-padding-top small-padding-right small-padding-left small-padding-bottom"
        >
          Name
        </th>
        <th
          className="cell-border text-align-left small-padding-top small-padding-right small-padding-left small-padding-bottom"
        >
          UserName
        </th>
        <th
          className="cell-border text-align-left small-padding-top small-padding-right small-padding-left small-padding-bottom"
        >
          E-mail
        </th>
        <th
          className="cell-border text-align-left small-padding-top small-padding-right small-padding-left small-padding-bottom"
        >
          Address
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
